File: 0146-lru-cache/0146-lru-cache.py
class Node:
    def __init__(self, key: int = 0, val: int = 0):
        self.key = key
        self.val = val
        self.prev = None
        self.next = None


class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.cache = {}  # key -> Node

        # Dummy head and tail to simplify boundary conditions
        self.head = Node()
        self.tail = Node()
        self.head.next = self.tail
        self.tail.prev = self.head

    def _remove(self, node: Node) -> None:
        """Remove an existing node from the linked list."""
        node.prev.next = node.next
        node.next.prev = node.prev

    def _add_first(self, node: Node) -> None:
        """Insert a node right after the dummy head (mark as most recently used)."""
        node.next = self.head.next
        node.prev = self.head
        self.head.next.prev = node
        self.head.next = node

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1

        node = self.cache[key]
        self._remove(node)
        self._add_first(node)
        return node.val

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            node = self.cache[key]
            node.val = value
            self._remove(node)
            self._add_first(node)
        else:
            new_node = Node(key, value)
            self.cache[key] = new_node
            self._add_first(new_node)

            if len(self.cache) > self.capacity:
                # Evict the least recently used node (node right before tail)
                lru = self.tail.prev
                self._remove(lru)
                del self.cache[lru.key]

File: 0146-lru-cache/0146-lru-cache.java
import java.util.HashMap;
import java.util.Map;

class LRUCache {

    private static class Node {
        int key;
        int value;
        Node prev;
        Node next;

        Node(int key, int value) {
            this.key = key;
            this.value = value;
        }
    }

    private final int capacity;
    private final Map<Integer, Node> cache;
    private final Node head;
    private final Node tail;

    public LRUCache(int capacity) {
        this.capacity = capacity;
        this.cache = new HashMap<>();

        // Initialize dummy head and tail
        this.head = new Node(0, 0);
        this.tail = new Node(0, 0);
        this.head.next = this.tail;
        this.tail.prev = this.head;
    }

    public int get(int key) {
        Node node = cache.get(key);
        if (node == null) {
            return -1;
        }
        // Move the accessed node to the front (most recently used)
        moveToHead(node);
        return node.value;
    }

    public void put(int key, int value) {
        Node node = cache.get(key);

        if (node != null) {
            // Key exists: update value and move to head
            node.value = value;
            moveToHead(node);
        } else {
            // Key does not exist: create and insert
            Node newNode = new Node(key, value);
            cache.put(key, newNode);
            addFirst(newNode);

            // Evict the least recently used element if over capacity
            if (cache.size() > capacity) {
                Node lru = removeTail();
                cache.remove(lru.key);
            }
        }
    }

    // Helper: Remove a node from the doubly-linked list
    private void removeNode(Node node) {
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }

    // Helper: Insert a node right after the dummy head
    private void addFirst(Node node) {
        node.next = head.next;
        node.prev = head;
        head.next.prev = node;
        head.next = node;
    }

    // Helper: Move an existing node to head
    private void moveToHead(Node node) {
        removeNode(node);
        addFirst(node);
    }

    // Helper: Pop the least recently used node (before tail)
    private Node removeTail() {
        Node res = tail.prev;
        removeNode(res);
        return res;
    }
}